builder/googlecompute/step_start_tunnel.go

```diff
@@ -0,0 +1,323 @@
+	// - You must manually enable the IAP API in the Google Cloud console.
+	// - You must have the gcloud sdk installed on the computer running Packer.
+	// - You must be using a Service Account with a credentials file (using the
+	//	 account_file option in the Packer template)
+	// - This is currently only implemented for the SSH communicator, not the
+	//   WinRM Communicator.
+	// - You must add the given service account to project level IAP permissions
+	//   in https://console.cloud.google.com/security/iap. To do so, click
+	//   "project" > "SSH and TCP resoures" > "All Tunnel Resources" >
+	//   "Add Member". Then add your service account and choose the role
+	//   "IAP-secured Tunnel User" and add any conditions you may care about.
+	IAP bool `mapstructure:"use_iap" required:"false"`
+	// Which port to connect the local end of the IAM localhost proxy to. If
+	// left blank, Packer will choose a port for you from available ports.
+	IAPLocalhostPort int `mapstructure:"iap_localhost_port"`
+	// What "hashbang" to use to invoke script that sets up gcloud.
+	// Default: "/bin/sh"
+	IAPHashBang string `mapstructure:"iap_hashbang" required:"false"`
+	// What file extension to use for script that sets up gcloud.
+	// Default: ".sh"
+	IAPExt string `mapstructure:"iap_ext" required:"false"`
+}
+
+type TunnelDriver interface {
+	StartTunnel(context.Context, string) error
+	StopTunnel()
+}
+
+func RunTunnelCommand(cmd *exec.Cmd) error {
+	// set stdout and stderr so we can read what's going on.
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Start()
+	if err != nil {
+		err := fmt.Errorf("Error calling gcloud sdk to launch IAP tunnel: %s",
+			err)
+		return err
+	}
+
+	// Give tunnel 30 seconds to either launch, or return an error.
+	// Unfortunately, the SDK doesn't provide any official acknowledgment that
+	// the tunnel is launched when it's not being run through a TTY so we
+	// are just trusting here that 30s is enough to know whether the tunnel
+	// launch was going to fail. Yep, feels icky to me too. But I spent an
+	// afternoon trying to figure out how to get the SDK to actually send
+	// the "Listening on port [n]" line I see when I run it manually, and I
+	// can't justify spending more time than that on aesthetics.
+	for i := 0; i < 30; i++ {
+		time.Sleep(1 * time.Second)
+
+		lineStderr, err := stderr.ReadString('\n')
+		if err != nil && err != io.EOF {
+			log.Printf("Err from scanning stderr is %s", err)
+			return fmt.Errorf("Error reading stderr from tunnel launch: %s", err)
+		}
+		if lineStderr != "" {
+			log.Printf("stderr: %s", lineStderr)
+		}
+
+		lineStdout, err := stdout.ReadString('\n')
+		if err != nil && err != io.EOF {
+			log.Printf("Err from scanning stdout is %s", err)
+			return fmt.Errorf("Error reading stdout from tunnel launch: %s", err)
+		}
+		if lineStdout != "" {
+			log.Printf("stdout: %s", lineStdout)
+		}
+
+		if strings.Contains(lineStderr, "ERROR") {
```

Comment:
I’m glad that the gcloud cli has some consistency in the error messaging. This is much better than searching for panic. On the AWS side I went with the idea of logging everything and if it fails then the ssh connection will yell loudly and timeout.

Reply:
Same. Frustrating that they don't return a success message, but the error messages are useful.